### PR TITLE
RomFS Extraction

### PR DESCRIPTION
**Pull request by DarkLordZach**
Converts a RomFS from a single binary blob to a `VirtualDir`.

Part of an upcoming suprise, figured I'd break it up.

### CODE REVIEW

File: src/common/file_util.cpp [hunk 1]
```diff
@@ -805,7 +805,7 @@ std::vector<std::string> SplitPathComponents(const std::string& filename) {
     std::replace(copy.begin(), copy.end(), '\\', '/');
     std::vector<std::string> out;
 
-    std::stringstream stream(filename);
+    std::stringstream stream(copy);
```
Comment (Subv) [inline src/common/file_util.cpp]: was this a bug?

File: src/core/file_sys/romfs.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    u32_le name_length;
+};
+static_assert(sizeof(FileEntry) == 0x20, "FileEntry has incorrect size.");
+
+template <typename Entry>
+static std::pair<Entry, std::string> GetEntry(const VirtualFile& file, size_t offset) {
+    Entry entry{};
+    if (file->ReadObject(&entry, offset) != sizeof(Entry))
+        return {{}, ""};
+    std::vector<char> string(entry.name_length);
```
Comment (Subv) [inline src/core/file_sys/romfs.cpp]: you can use std::string and use &string[0] instead of .data()

File: src/core/file_sys/vfs.h [hunk 3]
```diff
@@ -234,4 +246,7 @@ struct ReadOnlyVfsDirectory : public VfsDirectory {
     bool DeleteFile(const std::string& name) override;
     bool Rename(const std::string& name) override;
 };
+
+bool VfsRawCopy(VirtualFile src, VirtualFile dest);
```
Comment (Subv) [inline src/core/file_sys/vfs.h]: Add documentation for this function

File: src/core/file_sys/vfs_vector.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+template <typename T>
+static bool FindAndRemoveVectorElement(std::vector<T>& vec, const std::string& name) {
+    auto iter = std::find_if(vec.begin(), vec.end(), [name](T e) { return e->GetName() == name; });
+    if (iter == vec.end())
+        return false;
+    auto old_size = vec.size();
+    vec.erase(iter);
+    return old_size - 1 == vec.size();
```
Comment (Subv) [inline src/core/file_sys/vfs_vector.cpp]: isn't this always true?

File: src/common/file_util.cpp (hunk 1 above)
Comment (DarkLordZach) [reply src/common/file_util.cpp]: yes

File: src/core/file_sys/romfs.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    u32_le hash;
+    u32_le name_length;
+};
+static_assert(sizeof(FileEntry) == 0x20, "FileEntry has incorrect size.");
+
+template <typename Entry>
+static std::pair<Entry, std::string> GetEntry(const VirtualFile& file, size_t offset) {
+    Entry entry{};
+    if (file->ReadObject(&entry, offset) != sizeof(Entry))
+        return {{}, ""};
```
Comment (lioncash) [inline src/core/file_sys/romfs.cpp]: I'm fairly sure it's allowable to just do `return {};`

File: src/core/file_sys/romfs.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    std::string string(entry.name_length, 0);
```
Comment (lioncash) [inline src/core/file_sys/romfs.cpp]: `'\0'`

File: src/core/file_sys/romfs.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    const u64 file_offset = header.file_meta.offset;
+    const u64 dir_offset = header.directory_meta.offset + 4;
+
+    const auto root =
+        std::make_shared<VectorVfsDirectory>(std::vector<VirtualFile>{}, std::vector<VirtualDir>{},
+                                             file->GetContainingDirectory(), file->GetName());
+
+    ProcessDirectory(file, dir_offset, file_offset, 0, root);
+
+    VirtualDir out = root;
```
Comment (lioncash) [inline src/core/file_sys/romfs.cpp]: std::move(root);

File: src/core/file_sys/romfs.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2018 yuzu emulator team
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "common/common_funcs.h"
```
Comment (lioncash) [inline src/core/file_sys/romfs.h]: `common_types.h`. `<array>` also needs to be included.

File: src/core/file_sys/romfs.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    u32_le magic;
+    u32_le magic_number;
+    INSERT_PADDING_BYTES(8);
+    std::array<IVFCLevel, 6> levels;
+    INSERT_PADDING_BYTES(64);
+};
+static_assert(sizeof(IVFCHeader) == 0xE0, "IVFCHeader has incorrect size.");
+
+struct RomFSHeader;
+struct DirectoryEntry;
```
Comment (lioncash) [inline src/core/file_sys/romfs.h]: Why are these forward declarations in the header if nothing in it use them?

File: src/core/file_sys/vfs.cpp [hunk 10]
```diff
@@ -208,6 +215,13 @@ bool VfsDirectory::Copy(const std::string& src, const std::string& dest) {
     return f2->WriteBytes(f1->ReadAllBytes()) == f1->GetSize();
 }
 
+std::string VfsDirectory::GetFullPath() {
```
Comment (lioncash) [inline src/core/file_sys/vfs.cpp]: This can likely be a const member function.

File: src/core/file_sys/vfs.cpp [hunk 11]
```diff
@@ -46,6 +46,13 @@ size_t VfsFile::WriteBytes(std::vector<u8> data, size_t offset) {
     return Write(data.data(), data.size(), offset);
 }
 
+std::string VfsFile::GetFullPath() {
```
Comment (lioncash) [inline src/core/file_sys/vfs.cpp]: This can likely be a const member function.

File: src/core/file_sys/vfs_vector.cpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2018 yuzu emulator team
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "core/file_sys/vfs_vector.h"
```
Comment (lioncash) [inline src/core/file_sys/vfs_vector.cpp]: `<algorithm>` needs to be included.

File: src/core/file_sys/vfs_vector.h [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+#pragma once
+
+#include "core/file_sys/vfs.h"
+
+namespace FileSys {
+
+// An implementation of VfsDirectory that maintains two vectors for subdirectories and files.
+// Vector data is supplied upon construction.
+struct VectorVfsDirectory : public VfsDirectory {
+    VectorVfsDirectory(std::vector<VirtualFile> files = {}, std::vector<VirtualDir> dirs = {},
```
Comment (lioncash) [inline src/core/file_sys/vfs_vector.h]: `explicit VectorVfsDirectory`

File: src/core/file_sys/vfs_vector.cpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+std::shared_ptr<VfsDirectory> VectorVfsDirectory::CreateSubdirectory(const std::string& name) {
+    return nullptr;
+}
+
+std::shared_ptr<VfsFile> VectorVfsDirectory::CreateFile(const std::string& name) {
+    return nullptr;
+}
+
+void VectorVfsDirectory::AddFile(const VirtualFile& file) {
```
Comment (lioncash) [inline src/core/file_sys/vfs_vector.cpp]: This should likely take the parameter by value, and std::move it in a push_back.

File: src/core/file_sys/vfs_vector.cpp [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    files.emplace_back(file);
+}
+
+void VectorVfsDirectory::AddDirectory(const VirtualDir& dir) {
```
Comment (lioncash) [inline src/core/file_sys/vfs_vector.cpp]: Ditto

File: src/core/file_sys/romfs.h (hunk 8 above)
Comment (DarkLordZach) [reply src/core/file_sys/romfs.h]: insert_padding_bytes

**Pull request merged**